### PR TITLE
python: make every view a `venv`

### PR DESCRIPTION
**Pull request by tgamblin**
#40773 introduced `python-venv`, which improved build isolation and avoids issues with, e.g., `ubuntu`'s system python modifying `sysconfig` to include a (very unwanted) `local` directory within the default install layout.

This addresses a few cases where #40773 removed functionality, without harming the default cases where we use `python-venv`.

Traditionally, *every* view with `python` in it was essentially a virtual environment, because we would copy the `python` interpreter and `os.py` into every view when linking. We now rely on `python-venv` to do that, but only when it's used (i.e. new builds) and only for packages that have an `extends("python")` directive.

This again makes every view with `python` in it a virtual environment, but only if we're not already using a package like `python-venv`. This uses a different mechanism from before -- instead of using the `virtualenv` trick of copying `python` into the prefix, we instead create a `pyvenv.cfg` like `venv` (the more modern way to do it).

This fixes two things:
1. If you already had an environment before Spack `v0.22` that worked, it would stop working without a reconcretize and rebuild in `v0.22`, because we no longer copy the python interpreter on link. Adding `pyvenv.cfg` fixes this in a more modern way, so old views will keep working.

2. If you have an env that only includes python packages that use `depends_on("python")` instead of `extends("python")`, those packages will now be importable as before, though they won't have the same level of build isolation you'd get with `extends` and `python-venv`.

- [x] factor `python` finding logic out into its own method in `python/package.py`
- [x] add a method to make a `pyvenv.cfg`
- [x] have `add_files_to_view` create a `pyvenv.cfg` if none exists.

### CODE REVIEW

File: var/spack/repos/builtin/packages/python/package.py [hunk 1]
```diff
@@ -1242,6 +1267,35 @@ def setup_dependent_package(self, module, dependent_spec):
...
+
+        # location of python inside the view, where we will put the venv config
+        projection = view.get_projection_for_spec(self.spec)
+        pyvenv_cfg = os.path.join(projection, "pyvenv.cfg")
+        if os.path.lexists(pyvenv_cfg):
+            return
+
+        # if this is a copy view, the realpath of python will be in the view
+        python_path = find_python_in_prefix(self.spec, projection).path
+        copy_view = not os.path.commonpath((os.path.realpath(python_path), projection))
```
Comment (haampie) [inline var/spack/repos/builtin/packages/python/package.py]: If you use realpath you should use it on both lhs an rhs, otherwise it breaks when the view itself has a symlinked parent dir.

Also `not os.path.commonpath(...)` is always false since it is at least `/`.

Maybe expose the view type from the view object instead of inferring it?

File: var/spack/repos/builtin/packages/python/package.py [hunk 2]
```diff
@@ -1242,6 +1267,35 @@ def setup_dependent_package(self, module, dependent_spec):
...
+        super().add_files_to_view(view, merge_map, skip_if_exists=skip_if_exists)
+
+        # location of python inside the view, where we will put the venv config
+        projection = view.get_projection_for_spec(self.spec)
+        pyvenv_cfg = os.path.join(projection, "pyvenv.cfg")
+        if os.path.lexists(pyvenv_cfg):
+            return
+
+        # if this is a copy view, the realpath of python will be in the view
+        python_path = find_python_in_prefix(self.spec, projection).path
```
Comment (haampie) [inline var/spack/repos/builtin/packages/python/package.py]: Not a suggestion, but what I'm thinking reading this: I hope there's a way that relies less on implementation details, but don't immediately see it.

Current implementation relies on the order in which things are linked, and would break if we run it in parallel or so. The view code right now does: (1) collect what package links what, (2) create dir structure, and (3) link it. Step (3) is not order specific, so it could be done in parallel in the future.

Basically you'd just want to branch on whether `./pyvenv.cfg` is linked from any other package in this view (whether python-venv or other).

File: var/spack/repos/builtin/packages/python/package.py (hunk 1 above)
Comment (tgamblin) [reply var/spack/repos/builtin/packages/python/package.py]: Sigh. I wanted to but tried to avoid it b/c it is a bigger change. I think it's the best thing to do though.

File: var/spack/repos/builtin/packages/python/package.py (hunk 2 above)
Comment (tgamblin) [reply var/spack/repos/builtin/packages/python/package.py]: Right now we link sequentially, in a pretty sensible order. But yes it would be nice to be able to parallelize that. I don't see a way without rearchitecting how merge maps are done, though.

For this one you really want to know whether some other package will link a particular file into a view. I think the package that wants to know (here, python) would need to say so in advance and then when you construct the merge map you can fill in additional details to be passed in to `add_files_to_view`.  I don't see any other way to do it without computing everything in advance and passing all that to every merge... which seems like too much (and also likely limits parallelism).

File: var/spack/repos/builtin/packages/python/package.py (hunk 1 above)
Comment (tgamblin) [reply var/spack/repos/builtin/packages/python/package.py]: ok this is done.

**Review by haampie — APPROVED** [review]

**Review by haampie — APPROVED** [review]

**Comment by haampie** [comment]
@spackbot run pipeline
